### PR TITLE
feat: kerberos authenticcation

### PR DESCRIPTION
**Pull request by lafirest**
Fixes [EMQX-11630](https://emqx.atlassian.net/browse/EMQX-11630)

Release version: v/e5.8.0

## PR Checklist
Please convert it to a draft if any of the following conditions are not met. Reviewers may skip over until all the items are checked:

- [ ] Added tests for the changes
- [ ] Added property-based tests for code which performs user input validation
- [ ] Changed lines covered in coverage report
- [ ] Change log has been added to `changes/(ce|ee)/(feat|perf|fix|breaking)-<PR-id>.en.md` files
- [ ] For internal contributor: there is a jira ticket to track this change
- [ ] Created PR to [emqx-docs](https://github.com/emqx/emqx-docs) if documentation update is required, or link to a follow-up jira ticket
- [ ] Schema changes are backward compatible

## Checklist for CI (.github/workflows) changes

- [ ] If changed package build workflow, pass [this action](https://github.com/emqx/emqx/actions/workflows/build_packages.yaml) (manual trigger)
- [ ] Change log has been added to `changes/` dir for user-facing artifacts update

[EMQX-11630]: https://emqx.atlassian.net/browse/EMQX-11630?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: apps/emqx_auth_kerberos/src/emqx_auth_kerberos.app.src [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+    {applications, [
+        kernel,
+        stdlib,
+        emqx_auth,
+        sasl_auth
+    ]},
+    {env, []},
+    {modules, []},
+
+    {licenses, ["Apache 2.0"]},
```
Comment (zmstone) [inline apps/emqx_auth_kerberos/src/emqx_auth_kerberos.app.src:16]:
```suggestion
```

File: mix.exs [hunk 2]
```diff
@@ -376,9 +376,11 @@ defmodule EMQXUmbrella.MixProject do
       :emqx_gateway_jt808,
       :emqx_bridge_syskeeper,
       :emqx_ds_shared_sub,
+      :emqx_auth_gssapi,
```
Comment (zmstone) [inline mix.exs:379]:
```suggestion
```

File: .ci/docker-compose-file/docker-compose-kdc.yaml [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+  kdc:
+    hostname: kdc.emqx.net
+    image:  ghcr.io/emqx/emqx-builder/5.3-9:1.15.7-26.2.5-3-ubuntu22.04
+    container_name: kdc.emqx.net
+    expose:
+      - 88 # kdc
+      - 749 # admin server
+    # ports:
+    #   - 88:88
+    #   - 749:749
```
Comment (zmstone) [inline .ci/docker-compose-file/docker-compose-kdc.yaml:13]:
```suggestion
    #   - "88:88"
    #   - "749:749"
```

File: .ci/docker-compose-file/kerberos/run.sh [hunk 4]
```diff
@@ -6,20 +6,31 @@ echo "Remove old keytabs"
...
 kdb5_util -P emqx -r KDC.EMQX.NET create -s
 
 echo "Add principals"
 
-kadmin.local -w password -q "add_principal -randkey kafka/kafka-1.emqx.net@KDC.EMQX.NET"
+kadmin.local -w password -q "add_principal -randkey kafka/kafka-1.emqx.net@KDC.EMQX.NET" > /dev/null
 kadmin.local -w password -q "add_principal -randkey rig@KDC.EMQX.NET"  > /dev/null
 
+# For Kerberos Authn
+kadmin.local -w password -q "add_principal -randkey emqx/erlang.emqx.net@KDC.EMQX.NET" > /dev/null
```
Comment (zmstone) [inline .ci/docker-compose-file/kerberos/run.sh:22]:
```suggestion
kadmin.local -w password -q "add_principal -randkey mqtt/erlang.emqx.net@KDC.EMQX.NET" > /dev/null
```

File: .ci/docker-compose-file/kerberos/run.sh [hunk 5]
```diff
@@ -6,20 +6,31 @@ echo "Remove old keytabs"
...
+kadmin.local -w password -q "add_principal -randkey krb_authn_cli@KDC.EMQX.NET"  > /dev/null
+
 
 echo "Create keytabs"
 
 kadmin.local -w password -q "ktadd  -k /var/lib/secret/kafka.keytab -norandkey kafka/kafka-1.emqx.net@KDC.EMQX.NET " > /dev/null
 kadmin.local -w password -q "ktadd  -k /var/lib/secret/rig.keytab -norandkey rig@KDC.EMQX.NET " > /dev/null
 
+# For Kerberos Authn
+kadmin.local -w password -q "ktadd  -k /var/lib/secret/erlang.keytab -norandkey emqx/erlang.emqx.net@KDC.EMQX.NET " > /dev/null
```
Comment (zmstone) [inline .ci/docker-compose-file/kerberos/run.sh:32]:
```suggestion
kadmin.local -w password -q "ktadd  -k /var/lib/secret/erlang.keytab -norandkey mqtt/erlang.emqx.net@KDC.EMQX.NET " > /dev/null
```

File: apps/emqx_auth_kerberos/test/emqx_authn_kerberos_SUITE.erl [hunk 6]
```diff
@@ -0,0 +1,282 @@
...
+
+-include_lib("emqx/include/emqx_mqtt.hrl").
+-include_lib("emqx_auth/include/emqx_authn.hrl").
+
+-define(PATH, [authentication]).
+
+-define(INVALID_SVR_PRINCIPAL, <<"not-exists/erlang.emqx.nett@KDC.EMQX.NET">>).
+
+-define(SVR_HOST, "erlang.emqx.net").
+-define(SVR_PRINCIPAL, <<"emqx/erlang.emqx.net@KDC.EMQX.NET">>).
```
Comment (zmstone) [inline apps/emqx_auth_kerberos/test/emqx_authn_kerberos_SUITE.erl:22]:
```suggestion
-define(SVR_PRINCIPAL, <<"mqtt/erlang.emqx.net@KDC.EMQX.NET">>).
```

File: apps/emqx_auth_kerberos/test/emqx_authn_kerberos_SUITE.erl [hunk 7]
```diff
@@ -0,0 +1,282 @@
...
+    },
+    {continue, {?RC_CONTINUE_AUTHENTICATION, OutProps}, AuthState};
+do_auth_handle(_AuthState, _Reason, _Props) ->
+    {stop, protocol_error}.
+
+%%------------------------------------------------------------------------------
+%% Client Agent
+
+setup_cli() ->
+    Pid = erlang:spawn(fun() -> cli_agent_loop(#{}) end),
```
Comment (zmstone) [inline apps/emqx_auth_kerberos/test/emqx_authn_kerberos_SUITE.erl:248]: if we change emqtt to support init args, this spawn is not necessary.

**Comment by emqxqa** [comment]
[TestExecution](https://emqx.atlassian.net/browse/QA-3493)

File: apps/emqx_auth_kerberos/src/emqx_authn_kerberos.erl [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+
+create(
+    AuthenticatorID,
+    #{
+        principal := Principal,
+        keytab_file := KeyTabFile
+    }
+) ->
+    KeyTabPath = emqx_schema:naive_env_interpolation(KeyTabFile),
+    case sasl_auth:kinit(KeyTabPath, Principal) of
```
Comment (zmstone) [inline apps/emqx_auth_kerberos/src/emqx_authn_kerberos.erl:28]: if we force server to configure default keytab file, there is no need for server to do kinit.

Comment (zmstone) [reply apps/emqx_auth_kerberos/src/emqx_authn_kerberos.erl:31]: we may want to keep it here so errors can be detected sooner if principal is configured wrong for example.

**Review by zmstone — APPROVED** [review]

**Pull request merged**